simulation_server/virtual_accelerator/virtual_accelerator.py

```diff
@@ -0,0 +1,227 @@
+        every element for use in calculating the magnetic rigidity.
+        
+        Note: need to track on a copy of the lattice to not influence readings!
+        """
+        test_beam = ParticleBeam(
+            torch.zeros(1, 7), energy=self.initial_beam_distribution.energy
+        )
+        test_lattice = deepcopy(self.lattice)
+        element_names = [e.name for e in test_lattice.elements]
+        return dict(
+            zip(
+                element_names,
+                test_lattice.get_beam_attrs_along_segment(("energy",), test_beam)[0],
+            )
+        )
+
+    def set_shutter(self, value: bool):
+        """
+        Set the beam shutter state in the virtual accelerator simulator.
+        If `value` is True, the shutter is closed (no beam), otherwise it is open (beam present).
+        """
+        if value:
+            self.initial_beam_distribution.particle_charges = torch.tensor(0.0)
+        else:
+            self.initial_beam_distribution.particle_charges = (
+                self.initial_beam_distribution_charge
+            )
+
+        # run the simulation to update readings
+        self.lattice.track(incoming=self.initial_beam_distribution)
+
+    def set_pvs(self, values: dict):
+        """
+        Set the corresponding process variable (PV) to the given value on the virtual accelerator simulator.
+        """
+        for pv_name, value in values.items():
+            # handle the beam shutter separately
+            if pv_name == self.beam_shutter_pv:
+                self.set_shutter(value)
+                continue
+
+            if pv_name == "VIRT:BEAM:RESET_SIM":
+                self.reset()
+                continue
+
+            # get the base pv name
+            base_pv_name = ":".join(pv_name.split(":")[:3])
+            attribute_name = ":".join(pv_name.split(":")[3:])
+
+            # get the beam energy along the lattice -- returns a dict of element names to energies
+            beam_energy_along_lattice = self.beam_energy_along_lattice
+
+            # check if the pv_name is a control variable
+            if base_pv_name in self.mapping:
+                # set the value in the virtual accelerator simulator
+                element = getattr(self.lattice, self.mapping[base_pv_name].lower())
+
+                # get the beam energy for the element
+                energy = beam_energy_along_lattice[self.mapping[base_pv_name].lower()]
+
+                # if there are duplicate elements, just grab the first one (both will be adjusted)
+                if isinstance(element, list):
+                    element = element[0]
+
+                try:
+                    print("accessing element " + element.name + " to set PV " + pv_name + " to " + str(value))
+                    access_cheetah_attribute(element, attribute_name, energy, value)
+                except ValueError as e:
+                    raise ValueError(f"Failed to set PV {pv_name}: {str(e)}") from e
```

Comment:
This exception ends up bubbling up and causing the application to terminate. It should probably be handled at the write() method or post() callback level where it can be used to reject the write.